File: app/src/main/java/io/supertokens/session/SuperTokensHttpConnection.java
package io.supertokens.session;

import android.os.Build;
import androidx.annotation.RequiresApi;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.*;
import java.security.Permission;
import java.util.*;

public class SuperTokensHttpConnection extends HttpURLConnection {
    private HttpURLConnection connection;
    private ArrayList<Function> connectionOperations = new ArrayList<Function>();

    SuperTokensHttpConnection(URL url) throws IOException {
        super(url);
        connection =(HttpURLConnection) url.openConnection();
    }

    // URLConnection methods

    @Override
    public void connect() throws IOException {

    }

    public void setConnectTimeout(int timeout) {
        final int finalTimeout = timeout;
        connectionOperations.add(new Function() {
            @Override
            public void doBody(HttpURLConnection con) {
                con.setConnectTimeout(finalTimeout);
            }
        });
        connection.setConnectTimeout(finalTimeout);
    }

    public int getConnectTimeout() {
        return connection.getConnectTimeout();
    }

    public void setReadTimeout(int timeout) {
        connection.setReadTimeout(timeout);
    }

    public int getReadTimeout() {
        return connection.getReadTimeout();
    }

    public URL getURL() {
        return connection.getURL();
    }

    public int getContentLength() {
        return connection.getContentLength();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public long getContentLengthLong() {
        return connection.getContentLengthLong();
    }

    public String getContentType() {
        return connection.getContentType();
    }

    public String getContentEncoding() {
        return connection.getContentEncoding();
    }

    public long getExpiration() {
        return connection.getExpiration();
    }

    public long getDate() {
        return connection.getDate();
    }

    public long getLastModified() {
        return connection.getLastModified();
    }

    public String getHeaderField(String name) {
        return connection.getHeaderField(name);
    }

    public Map<String, List<String>> getHeaderFields() {
        return connection.getHeaderFields();
    }

    public int getHeaderFieldInt(String name, int Default) {
        return connection.getHeaderFieldInt(name, Default);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public long getHeaderFieldLong(String name, long Default) {
        return connection.getHeaderFieldLong(name, Default);
    }

    @Override
    public void disconnect() {

    }

    @Override
    public boolean usingProxy() {
        return false;
    }

    public Object getContent() throws IOException {
        return connection.getContent();
    }

    public Object getContent(Class[] classes) throws IOException {
        return connection.getContent(classes);
    }


    public InputStream getInputStream() throws IOException {
        return connection.getInputStream();
    }

    public OutputStream getOutputStream() throws IOException {
        return connection.getOutputStream();
    }

    public String toString() {
        return connection.toString();
    }

    public void setDoInput(boolean doinput) {
        connection.setDoInput(doinput);
    }

    public boolean getDoInput() {
        return connection.getDoInput();
    }

    public void setDoOutput(boolean dooutput) {
        connection.setDoOutput(dooutput);
    }

    public boolean getDoOutput() {
        return connection.getDoOutput();
    }

    public void setAllowUserInteraction(boolean allowuserinteraction) {
        connection.setAllowUserInteraction(allowuserinteraction);
    }

    public boolean getAllowUserInteraction() {
        return connection.getAllowUserInteraction();
    }

    public void setUseCaches(boolean usecaches) {
        connection.setUseCaches(usecaches);
    }

    public boolean getUseCaches() {
        return connection.getUseCaches();
    }

    public void setIfModifiedSince(long ifmodifiedsince) {
        connection.setIfModifiedSince(ifmodifiedsince);
    }

    public long getIfModifiedSince() {
        return connection.getIfModifiedSince();
    }

    public boolean getDefaultUseCaches() {
        return connection.getDefaultUseCaches();
    }

    public void setDefaultUseCaches(boolean defaultusecaches) {
        connection.setDefaultUseCaches(defaultusecaches);
    }

    public void setRequestProperty(String key, String value) {
        connection.setRequestProperty(key, value);
    }

    public void addRequestProperty(String key, String value) {
        connection.addRequestProperty(key, value);
    }

    public String getRequestProperty(String key) {
        return  connection.getRequestProperty(key);
    }

    public Map<String,List<String>> getRequestProperties() {
        return connection.getRequestProperties();
    }

    // END URLConnection methods

    // HttpURLConnection methods

    /**
     * Returns the key for the {@code n}<sup>th</sup> header field.
     * Some implementations may treat the {@code 0}<sup>th</sup>
     * header field as special, i.e. as the status line returned by the HTTP
     * server. In this case, {@link #getHeaderField(int) getHeaderField(0)} returns the status
     * line, but {@code getHeaderFieldKey(0)} returns null.
     *
     * @param   n   an index, where {@code n >=0}.
     * @return  the key for the {@code n}<sup>th</sup> header field,
     *          or {@code null} if the key does not exist.
     */
    public String getHeaderFieldKey (int n) {
        return null;
    }

    public void setFixedLengthStreamingMode (int contentLength) {
        connection.setFixedLengthStreamingMode(contentLength);
    }

    public void setFixedLengthStreamingMode(long contentLength) {
        connection.setFixedLengthStreamingMode(contentLength);
    }

    public void setChunkedStreamingMode (int chunklen) {
        connection.setChunkedStreamingMode(chunklen);
    }

    public String getHeaderField(int n) {
        return null;
    }

    public void setInstanceFollowRedirects(boolean followRedirects) {
        connection.setInstanceFollowRedirects(followRedirects);
    }

    public boolean getInstanceFollowRedirects() {
        return connection.getInstanceFollowRedirects();
    }

    public void setRequestMethod(String method) throws ProtocolException {
        connection.setRequestMethod(method);
    }

    public String getRequestMethod() {
        return connection.getRequestMethod();
    }

    public int getResponseCode() throws IOException {
        return connection.getResponseCode();
    }

    public String getResponseMessage() throws IOException {
        return connection.getResponseMessage();
    }

    @SuppressWarnings("deprecation")
    public long getHeaderFieldDate(String name, long Default) {
        return connection.getHeaderFieldDate(name, Default);
    }

    public Permission getPermission() throws IOException {
        return connection.getPermission();
    }

    public InputStream getErrorStream() {
        return connection.getErrorStream();
    }


    private interface Function {
        void doBody(HttpURLConnection connection);
    }
}

File: app/src/main/java/io/supertokens/session/SuperTokensHTTPConnection.kt
package io.supertokens.session

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.*
import java.security.Permission

@Suppress("DeprecatedCallableAddReplaceWith")
class SuperTokensHttpURLConnection(url: String) {
    private val instance = URL(url).openConnection() as HttpURLConnection
    var readTimeout: Int
        get() = instance.readTimeout
        set(value) {
            instance.readTimeout = value
        }


    companion object {
        // UrlConnection methods
         @JvmStatic
         fun getFileNameMap(): FileNameMap {
             return URLConnection.getFileNameMap()
         }

        @JvmStatic
        fun setFileNameMap(map: FileNameMap) {
            URLConnection.setFileNameMap(map)
        }

        @JvmStatic
        fun setDefaultAllowUserInteraction(defaultallowuserinteraction: Boolean) {
            URLConnection.setDefaultAllowUserInteraction(defaultallowuserinteraction)
        }

        @JvmStatic
        fun getDefaultAllowUserInteraction(): Boolean {
            return URLConnection.getDefaultAllowUserInteraction()
        }

        @Deprecated("The instance specific setRequestProperty method should be used after an appropriate instance of URLConnection is obtained. Invoking this method will have no effect.")
        @JvmStatic
        fun setDefaultRequestProperty(key: String, value: String) {
            URLConnection.setDefaultRequestProperty(key, value)
        }

        @Deprecated("The instance specific getRequestProperty method should be used after an appropriate instance of URLConnection is obtained.")
        @JvmStatic
        fun getDefaultRequestProperty(key: String): String {
            return URLConnection.getDefaultRequestProperty(key)
        }

        @JvmStatic
        fun setContentHandlerFactory(fac: ContentHandlerFactory) {
            URLConnection.setContentHandlerFactory(fac)
        }

        @JvmStatic
        fun guessContentTypeFromName(fname: String): String {
            return URLConnection.guessContentTypeFromName(fname)
        }

        @Throws(IOException::class)
        @JvmStatic
        fun guessContentTypeFromStream(ins: InputStream): String? {
            return URLConnection.guessContentTypeFromStream(ins)
        }
        // END URLConnection Methods

        //HttpURLConnection methods

        @JvmStatic
        fun setFollowRedirects(set: Boolean) {
            HttpURLConnection.setFollowRedirects(set)
        }

        @JvmStatic
        fun getFollowRedirects(): Boolean {
            return HttpURLConnection.getFollowRedirects()
        }
    }

    // URLConnection methods
    fun setConnectTimeout(timeout: Int) {
        instance.connectTimeout = timeout
    }

    fun getConnectTimeout(): Int {
        return instance.connectTimeout
    }

//    fun setReadTimeout(timeout: Int) {
//        readTimeout = timeout
////        instance.readTimeout = timeout
//    }

//    fun getReadTimeout(): Int {
////        readTimeout = instance.readTimeout
//        return readTimeout
//    }

    fun getURL(): URL {
        return instance.url
    }

    fun getContentLength(): Int {
        return instance.contentLength
    }

    fun getContentLengthLong(): Long {
        if ( android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.N ) {
            return -1
        }
        return instance.contentLengthLong
    }

    fun getContentType(): String? {
        return instance.contentType
    }

    fun getContentEncoding(): String? {
        return instance.contentEncoding
    }

    fun getExpiration(): Long {
        return instance.expiration
    }

    fun getDate(): Long {
        return instance.date
    }

    fun getLastModified(): Long {
        return instance.lastModified
    }

    fun getHeaderField(name: String): String? {
        return instance.getHeaderField(name)
    }

    fun getHeaderFields(): Map<String, List<String>> {
        return instance.headerFields
    }

    fun getHeaderFieldInt(name: String, Default: Int): Int {
        return instance.getHeaderFieldInt(name, Default)
    }

    fun getHeaderFieldLong(name: String, Default: Long): Long {
        if ( android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.N ) {
            return -1
        }
        return instance.getHeaderFieldLong(name, Default)
    }

    fun getHeaderFieldKey(n: Int): String? {
        return instance.getHeaderFieldKey(n)
    }

    fun getHeaderField(n: Int): String? {
        return instance.getHeaderField(n)
    }

    @Throws(IOException::class)
    fun getContent(): Any {
        return instance.content
    }

    @Throws(IOException::class)
    fun getContent(classes: Array<Class<*>>): Any {
        return instance.getContent(classes)
    }

    @Throws(IOException::class)
    fun getInputStream(): InputStream {
        return instance.inputStream
    }

    @Throws(IOException::class)
    fun getOutputStream(): OutputStream {
        return instance.outputStream
    }

    override fun toString(): String {
        return instance.toString()
    }

    fun setDoInput(doinput: Boolean) {
        instance.doInput = doinput
    }

    fun getDoInput(): Boolean {
        return instance.doInput
    }

    fun setDoOutput(dooutput: Boolean) {
        instance.doOutput = dooutput
    }

    fun getDoOutput(): Boolean {
        return instance.doOutput
    }

    fun setAllowUserInteraction(allowuserinteraction: Boolean) {
        instance.allowUserInteraction = allowuserinteraction
    }

    fun getAllowUserInteraction(): Boolean {
        return instance.allowUserInteraction
    }

    fun setUseCaches(usecaches: Boolean) {
        instance.useCaches = usecaches
    }

    fun getUseCaches(): Boolean {
        return instance.useCaches
    }

    fun setIfModifiedSince(ifmodifiedsince: Long) {
        instance.ifModifiedSince = ifmodifiedsince
    }

    fun getIfModifiedSince(): Long {
        return instance.ifModifiedSince
    }

    fun getDefaultUseCaches(): Boolean {
        return instance.defaultUseCaches
    }

    fun setDefaultUseCaches(defaultusecaches: Boolean) {
        instance.defaultUseCaches = defaultusecaches
    }

    fun setRequestProperty(key: String?, value: String) {
        instance.setRequestProperty(key, value)
    }

    fun addRequestProperty(key: String?, value: String) {
        instance.addRequestProperty(key, value)
    }

    fun getRequestProperty(key: String): String? {
       return instance.getRequestProperty(key)
    }

    fun getRequestProperties(): Map<String, List<String>> {
        return instance.requestProperties

    }
    // END URLConnection methods

    // HttpUrlConnection methods

    fun setFixedLengthStreamingMode(contentLength: Int) {
        instance.setFixedLengthStreamingMode(contentLength)
    }

    fun setFixedLengthStreamingMode(contentLength: Long) {
        instance.setFixedLengthStreamingMode(contentLength)
    }

    fun setChunkedStreamingMode(chunklen: Int) {
        instance.setChunkedStreamingMode(chunklen)
    }

    fun setInstanceFollowRedirects(followRedirects: Boolean) {
        instance.instanceFollowRedirects = followRedirects
    }

    fun getInstanceFollowRedirects(): Boolean {
        return instance.instanceFollowRedirects
    }

    @Throws(ProtocolException::class)
    fun setRequestMethod(method: String) {
        instance.requestMethod = method
    }

    fun getRequestMethod(): String {
        return instance.requestMethod
    }

    @Throws(IOException::class)
    fun getResponseCode(): Int {
        return instance.responseCode
    }

    @Throws(IOException::class)
    fun getResponseMessage(): String {
        return instance.responseMessage
    }

    fun getHeaderFieldDate(name: String, Default: Long): Long {
        return instance.getHeaderFieldDate(name, Default)
    }

    @Throws(IOException::class)
    fun getPermission(): Permission {
        return instance.permission
    }

    fun getErrorStream(): InputStream? {
        return instance.errorStream
    }
 }